Sources/Private/Experimental/Animations/GradientAnimations.swift

```diff
@@ -0,0 +1,129 @@
+// Created by Cal Stephens on 1/7/22.
+// Copyright © 2022 Airbnb Inc. All rights reserved.
+
+import QuartzCore
+
+extension GradientRenderLayer {
+
+  // MARK: Internal
+
+  /// Adds gradient-related animations to this layer, from the given `GradientFill`
+  func addAnimations(for gradientFill: GradientFill, context: LayerAnimationContext) {
+    // We have to set `colors` to a non-nil value with some valid number of colors
+    // for the color animation below to have any effect
+    colors = .init(
+      repeating: CGColor.rgb(0, 0, 0),
+      count: gradientFill.numberOfColors)
+
+    addAnimation(
+      for: .colors,
+      keyframes: gradientFill.colors.keyframes,
+      value: { colorComponents in
+        gradientFill.colorConfiguration(from: colorComponents).map { $0.color }
+      },
+      context: context)
+
+    addAnimation(
+      for: .locations,
+      keyframes: gradientFill.colors.keyframes,
+      value: { colorComponents in
+        gradientFill.colorConfiguration(from: colorComponents).map { $0.location }
+      },
+      context: context)
+
+    switch gradientFill.gradientType {
+    case .linear:
+      addLinearGradientAnimations(for: gradientFill, context: context)
+    case .radial:
+      addRadialGradientAnimations(for: gradientFill, context: context)
+    case .none:
+      break
+    }
+  }
+
+  // MARK: Private
+
+  private func addLinearGradientAnimations(
+    for gradientFill: GradientFill,
+    context: LayerAnimationContext)
+  {
+    type = .axial
+
+    addAnimation(
+      for: .startPoint,
+      keyframes: gradientFill.startPoint.keyframes,
+      value: { absoluteStartPoint in
+        percentBasedPointInBounds(from: absoluteStartPoint.pointValue)
+      },
+      context: context)
+
+    addAnimation(
+      for: .endPoint,
+      keyframes: gradientFill.endPoint.keyframes,
+      value: { absoluteEndPoint in
+        percentBasedPointInBounds(from: absoluteEndPoint.pointValue)
+      },
+      context: context)
+  }
+
+  private func addRadialGradientAnimations(
+    for gradientFill: GradientFill,
+    context _: LayerAnimationContext)
+  {
+    type = .radial
+
+    // To draw the correct gradients, we have to derive a custom `endPoint`
+    // relative to the `startPoint` value. Since calculating the `endPoint`
+    // at any given time requires knowing the current `startPoint`,
+    // we can't allow them to animate separately.
+    guard
+      gradientFill.startPoint.keyframes.count == 1,
+      gradientFill.endPoint.keyframes.count == 1
+    else {
+      fatalError("Animating radial gradient `endPoint` point is currently unsupported")
```

Comment:
Seems like we could fail more gracefully here vs. crashing the host app on an unsupported feature—unless you're planning on addressing immediately in a follow-up?

Reply:
Good question -- I've been following this pattern elsewhere for now ([example](https://github.com/airbnb/lottie-ios/blob/master/Sources/Private/Experimental/Animations/EllipseAnimation.swift#L24). Relatively soon we're going to need to figure out a long term plan for unsupportable features (e.g. features that exist in the legacy rendering engine that just can't be expressed in a `CAKeyframeAnimation`). 

I have a feeling that the new engine can/will support enough features to become the default eventually, and we keep the legacy engine around to be used if necessary. Not totally sure though.

Reply:
I could imagine some sort of system where we make animation construction throwing and recover by falling back to the legacy main-thread renderer

Reply:
> I've been following this pattern elsewhere for now

As far as I can tell this pattern is explicitly against our [style guide rule](https://github.com/airbnb/swift#preconditions-and-asserts):

> Handle an unexpected but recoverable condition with an assert method combined with the appropriate logging in production. If the unexpected condition is not recoverable, prefer a precondition method or fatalError(). This strikes a balance between crashing and providing insight into unexpected conditions in the wild. Only prefer fatalError over a precondition method when the failure message is dynamic, since a precondition method won't report the message in the crash report.

Ideally we'd be able to clean up this type of behavior at all possible in cases where the errors are recoverable as a framework should never be crashing a consumer app if at all possible.

Reply:
> I've been following this pattern elsewhere for now

As far as I can tell this pattern is explicitly against our [style guide rule](https://github.com/airbnb/swift#preconditions-and-asserts):

> Handle an unexpected but recoverable condition with an assert method combined with the appropriate logging in production. If the unexpected condition is not recoverable, prefer a precondition method or fatalError(). This strikes a balance between crashing and providing insight into unexpected conditions in the wild. Only prefer fatalError over a precondition method when the failure message is dynamic, since a precondition method won't report the message in the crash report.

Ideally we'd be able to clean up this type of behavior at all possible in cases where the errors are recoverable as a framework should never be crashing a consumer app if at all possible.a

Reply:
> I could imagine some sort of system where we make animation construction throwing and recover by falling back to the legacy main-thread renderer

This is probably the best approach for making the new engine the default (We'll Get There™)

Agree that we shouldn't be `fatalError`ing here once this is used in production